packages/headless/src/features/insight-user-actions/insight-user-actions-preprocessing.ts

```diff
@@ -0,0 +1,274 @@
+} from './insight-user-actions-state';
+
+const MSEC_IN_SECOND = 1000;
+const SECONDS_IN_MINUTE = 60;
+const MAX_MINUTES_IN_SESSION = 30;
+const THIRTY_MINUTES =
+  MAX_MINUTES_IN_SESSION * SECONDS_IN_MINUTE * MSEC_IN_SECOND;
+
+interface rawUserAction {
+  name: string;
+  value: string;
+  time: string;
+}
+
+export const preprocessActionsData = (
+  state: UserActionsState,
+  actions: Array<rawUserAction>
+): UserActionTimeline => {
+  const ticketCreationDate = state.ticketCreationDate;
+  const excludedCustomActions = state.excludedCustomActions;
+
+  if (!ticketCreationDate || actions.length === 0) {
+    return {
+      precedingSessions: [],
+      session: undefined,
+      followingSessions: [],
+      caseSessionFound: false,
+    };
+  }
+
+  const processedRawActions = buildUserActionFromRawAction(actions);
+
+  const timeline = splitActionsIntoTimelineSessions(
+    processedRawActions,
+    ticketCreationDate
+  );
+
+  const filteredTimeline = filterTimelineActions(
+    timeline,
+    excludedCustomActions
+  );
+
+  return filteredTimeline;
+};
+
+export const buildUserActionFromRawAction = (
+  rawActions: rawUserAction[]
+): UserAction[] => {
+  const mappedUserActions = rawActions
+    .map((rawAction) => {
+      const actionData = JSON.parse(rawAction.value);
+      return {
+        actionType: rawAction.name as UserActionType,
+        timestamp: rawAction.time,
+        eventData: {
+          type: actionData.event_type,
+          value: actionData.event_value,
+        },
+        cause: actionData.cause,
+        searchHub: actionData.origin_level_1,
+        document: {
+          title: actionData.title,
+          uriHash: actionData.uri_hash,
+          // eslint-disable-next-line @cspell/spellchecker
+          contentIdKey: actionData.c_contentidkey,
+          // eslint-disable-next-line @cspell/spellchecker
+          contentIdValue: actionData.c_contentidvalue,
+        },
+        query: actionData.query_expression,
+      };
+    })
+    .sort((a, b) => Number(b.timestamp) - Number(a.timestamp));
+
+  return mappedUserActions;
+};
+
+/**
+ * Splits the actions into sessions and build the timeline.
+ * @param actions {UserAction[]} - The actions to split
+ * @param ticketCreationDate {string} - The ticket creation date
+ * @returns {UserActionTimeline}
+ */
+export const splitActionsIntoTimelineSessions = (
+  actions: UserAction[],
+  ticketCreationDate: string
+): UserActionTimeline => {
+  const returnTimeline: UserActionTimeline = {
+    precedingSessions: [],
+    session: undefined,
+    followingSessions: [],
+    caseSessionFound: false,
+  };
+
+  let currentSession: UserSession = {
+    start: actions[0].timestamp,
+    end: actions[0].timestamp,
+    actions: [],
+  };
+
+  actions.forEach((action) => {
+    if (isActionWithin30Minutes(action, currentSession.end)) {
+      currentSession.actions.push(action);
+      currentSession.start = action.timestamp;
+      return;
+    }
+
```

Comment:
At this point you know that we will be starting a new session so you wanna do the following:

1.  Insert the case creation event in the session if necessary.
2. Position the session correctly in the timeline: preceding session, following session or activeCaseCreation session.

You can directly here start with point number 1:
 
You can create a general utility function:
```
function isTimestampInRange(timestamp, startTimestamp, endTimestamp) {
    return timestamp >= startTimestamp && timestamp <= endTimestamp;
}
 ```

and then use it in line 112 by doing: 
```
      const caseCreationIsPartOfSession = isTimestampBetween(caseCreationDate, currentSession.start, currentSession.end)
      if(caseCreationIsPartOfSession){
        insertActionInSession(currentSession, {type: 'CUSTOM', value: 'CaseCreation', time: 123})
      }
```

then we address point number 2, which positioning the session in the timeline which can be a separate utility function:

```
positionSessionInTimeline(session, state)
```

which is a function that basically position the session based on its start/end timestamp in the preceding session or the following session or the activeCaseCreation session.